### DISCUSSION TITLE
ImportError: attempted relative import with no known parent package

### DISCUSSION THREAD

**Discussion by JFuqX** [dashingsoft/pyarmor#1130] [Q&A] [answered]
After trying to run my obfuscated project I get the following issue:

```
from .pyarmor_runtime_0019xx import __pyarmor__
ImportError: attempted relative import with no known parent package
```

**Comment by JFuqX** [comment]
<img width="170" alt="Bildschirmfoto 2023-04-11 um 17 18 40" src="https://user-images.githubusercontent.com/56825613/231210187-7662b7d8-8dfa-4f25-998f-a3c434be57a9.png">
<img width="459" alt="Bildschirmfoto 2023-04-11 um 17 18 46" src="https://user-images.githubusercontent.com/56825613/231210201-aa7c4006-2938-4ccd-87ab-f31842f7d4a4.png">

I mean somehow the imports aren't working this is why I can't run obfuscated scripts
but not sure why this is the case.

**Comment by jachinte** [comment]
@jondy  I'm having the same problem. Here's how I'm running pyarmor:

```commandline
pyarmor gen -r -i src/mypackage
```

Here are the logs:

```
INFO     Python 3.9.16
INFO     Pyarmor 8.1.5 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     search inputs ...
INFO     find package at src/mypackage
INFO     find 1 top resources
INFO     start to generate runtime files
INFO     target platforms {'linux.x86_64'}
INFO     write dist/mypackage/pyarmor_runtime_000000/pyarmor_runtime.so
INFO     patch runtime file
INFO     generate runtime files OK
INFO     start to obfuscate scripts
...
```

The dist package directory looks like this:

```
.
├── mypackage
│   ├── __init__.py
│   ├── __main__.py
│   ├── __pycache__
│   ├── __version__.py
│   ├── other dirs
...
```

And `__main__.py` looks like this:

```
cat __main__.py 
# Pyarmor 8.1.5 (trial), 000000, non-profits, 2023-04-14T15:19:19.713198
from .pyarmor_runtime_000000 import __pyarmor__
__pyarmor__(__name__, __file__, b'PY...')
```

This is what happens when I try to run `__main__.py`:

```
python __main__.py 
Traceback (most recent call last):
  File "/app/dist/mypackage/__main__.py", line 2, in <module>
    from .pyarmor_runtime_000000 import __pyarmor__
ImportError: attempted relative import with no known parent package
```

I'm considering buying a pro license, but I won't until I can confirm my software will run as expected.

**Reply by jachinte** [reply]
Just in case anyone is looking for a way to make this work:

```commandline
pyarmor gen -r src/mypackage
```

The reason why I was getting this error is because I was missing a `__main__.py` file in the `dist` directory for running mypackage. After creating one, I could successfully run:

```
python ./dist
```

**Comment by b3n3w** [comment]
Having the same problem with our libraries

**Comment by JFuqX** [comment]
I have bought already a pro license but support isn't the best rn ... so I wait until pyarmor 8 is operational

**Reply by pauloswear** [reply]
Did you fix it?

**Comment by jondy** [comment, marked as answer]
@JFuqX please check to solve this problem
http://pyarmor.readthedocs.io/en/stable/questions.html
http://pyarmor.readthedocs.io/en/stable/reference/errors.html

This is wrong usage, all of them could be solved by reading the documentation, I may not answer this kind of question.

**Reply by CapricornJay** [reply]
Horrible Documentation

**Reply by 1328649047** [reply]
> 糟糕的文档

yes

**Comment by xthyax** [comment]
It seems I'm late for the party, but I think I have figured out a way to work on this, in case for future users.
I have look into the docs, it seem that it focus on the package and script stuff, so if you want to obfuscasting the whole project, here is how can you do it.
Example the project may look like this:
<pre>
myproject
|- moduleA 
| |- scriptA.py
| |- scriptB.py
|- moduleB
| |- scriptC.py
| |- scriptD.py
|- main.py
</pre>
Instead of doing this:
`pyarmor gen -O encoded_project -r -i myproject`
Do this instead:
```python
import os
for module in module_list:
 os.system(f"pyarmor gen -O encoded_project -r -i {module}")
os.system("pyarmor gen -O encoded_project -i main.py")
```
The idea is that you should obfuscating the packages and scripts only instead of obfuscate the whole project.
If you have multiple packages, obfuscate them separately.
Hope this help.

**Comment by ChaunceyWang** [comment]
as the ducument mentioned, https://pyarmor.readthedocs.io/en/stable/reference/errors.html
![image](https://github.com/dashingsoft/pyarmor/assets/17997053/626ffff5-26ac-4e6a-bbf8-9031b30b52e4)

Do not use [-i](https://pyarmor.readthedocs.io/en/stable/reference/man.html#cmdoption-pyarmor-gen-i) or [--prefix](https://pyarmor.readthedocs.io/en/stable/reference/man.html#cmdoption-pyarmor-gen-prefix) if you don’t know what they’re doing.

### This works for me:

`pyarmor gen -O dist -r  my_folder`

The **my_folder** directory is:
```
├── A
│   ├── AA
│   │   ├── aa.py
│   ├── a.py
├── B
│   ├── b.py
└── main.py
```

in **dist** directory, I  get:
```
├── my_folder
└── pyarmor_runtime_000000
```
Then **(important)**

`mv  pyarmor_runtime_000000  my_folder ` 

go to **my_folder** directory，codes can be executed correctly

`python main.py`

**Reply by xthyax** [reply]
I guess ignoring the prefix `-i` seems to be a good choice. I'll give your recommendation a go.

But to be honest, the whole reasons is not really elaborate enough to me.

**Reply by ChaunceyWang** [reply]
agree, hope author @jondy fix this problem. Processing on the entire directory is friendly to use

**Comment by jondy** [comment]
Try to use both `-i` and `--prefix`, for example 
```
  src/
     main.py
     pkg1/
        __init__.py
        util.py
     pkg2/
        util.py
```

And obfuscate all of them by this command:
```
pyarmor gen -i --prefix pkg1 src/main.py src/pkg1 src/pkg2
```

Run the obfuscated script:
```
mv dist/ obf_src
cd obf_src
python main.py
```

Check the whole content of `obf_src` and the content of each `__init__.py` to understand the meaning `-i` and `--prefix`

If need not save runtime package in any child folder, it's easy by this way:
```
pyarmor gen -r src
mv dist obf_src
cd obf_src
python main.py
```
